### PR TITLE
WIP. Feature/configuration loader (contains smtp-service work)

### PR DESCRIPTION
**Pull request by spideynr**
I have some questions before proceeding with this one.

1) I made this util into a class because all of the utils were classes. But it can also be converted to a pure function since it only merges two objects.

2) I currently use the yeep.config.sample that you have written for documentation purposes as the default object. While this seems initially bad, having a _third_ configuration file that contains the defaults will mean we have to update two places with the sane, default, values whenever we add something, which i think is gonna lead to inconsistencies in the future. We can be more defensive over the sample file, maybe rename it and keep it in a different folder. Any objections with that?

3) The original idea to create the configurationLoader was to avoid doing unnecessary checks for options from the configuration files. Unfortunately the problem persists in the `server/events/passwordResetInit` as it tries to initialise outside the scope of the server setup. This means that the configuration loader needs to either be injected there as well, or we need to refactor files containing separate logic. Do you want to handle these initialisations in a different way? Maybe pass them through the server setup?

### CODE REVIEW

**Comment by jmike** [comment]
Great questions - let me attempt to answer them inline.

> I made this util into a class because all of the utils were classes. But it can also be converted to a pure function since it only merges two objects.

Agreed, let's turn it into a pure function instead - how does `loadConfig()` sound?
FYI you may also have functions in `server/utils` - it's not exclusive to classes.

> I currently use the yeep.config.sample that you have written for documentation purposes as the default object. While this seems initially bad, having a third configuration file that contains the defaults will mean we have to update two places with the sane, default, values whenever we add something, which i think is gonna lead to inconsistencies in the future. We can be more defensive over the sample file, maybe rename it and keep it in a different folder. Any objections with that?

We can remove the sample file altogether since we agreed to implement the `yeep init` CLI. The original idea was that the user would copy the sample file and edit it accordingly. However, `yeep init` is much better in terms of dev experience.

tl;dr feel free to remove it.

 > The original idea to create the configurationLoader was to avoid doing unnecessary checks for options from the configuration files. Unfortunately the problem persists in the server/events/passwordResetInit as it tries to initialise outside the scope of the server setup. This means that the configuration loader needs to either be injected there as well, or we need to refactor files containing separate logic. Do you want to handle these initialisations in a different way? Maybe pass them through the server setup?

I suggest we annotate/normalize the config file at load time and then pass it down to any interested party via our cool `ctx` object. So essentially we only load the config file once using our `loadConfig()` function. Makes sense?

**Comment by jmike** [comment]
Btw this PR looks great so far.

**Comment by spideynr** [comment]
Code is already merged to master. Closing this PR.
